README
1. Bio-CUA

Version: 1.01

2. Purpose

The aim of this distribution is to provide comprehensive and flexible
tools to analyze codon usage bias (CUB) and relevant problems.

One amino acid can be encoded by more than one synonymous codon, and
synonymous codons are unevenly used. For example, some codons are used
more often than other synonymous ones in highly expressed genes
(I<Sharp and Li 1987>). To measure the unevenness of codon usage, multiple
metrics of codon usage bias have been developed, such as Fop
(Frequency of optimal codons), CAI (Codon Adaptation Index), tAI (tRNA
Adaptation Index), and ENC (Effective Number of Codons). Biased codon
usage is widespread, visible in all species. It is important both to
identify codons having high translational efficiency (often termed
optimal codons) and to study the distribution of codon usage among
genes (e.g., genes with more optimal codons versus genes with fewer
optimal codons).

So far, no software implements all the above CUB metrics in one place.
More importantly, parameters in existing software are often fixed,
so one can compute certain types of CUB metrics for a limited list of 
species and cannot tune parameters. For example, when one wants to
identify optimal codons in certain tissues, it may be better to use
most highly expressed genes of that tissue to calculate CAI, which is
impossible with existing software.

This package mainly solves these two problems: providing tools
computing all common CUB metrics and allowing users to tune parameters
freely. We also incorporate or extend some method variants, such as 
GC-content corrected ENC, background-data normalized CAI, etc. 
See the relevant methods' description in CUB classes for more details.


3. INSTALLATION

To install this module, run the following commands:

a. If install after downloading the package,

	perl Makefile.PL
	make
	make test
	make install

b. install directly from CPAN as

	cpan Bio::CUA

4. SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Bio::CUA

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Bio-CUA

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Bio-CUA

    CPAN Ratings
        http://cpanratings.perl.org/d/Bio-CUA

    Search CPAN
        http://search.cpan.org/dist/Bio-CUA/


You can also email me at zhangz.sci@gmail.com for help.

5. LICENSE AND COPYRIGHT

Copyright (C) 2015 Zhenguo Zhang

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see L<http://www.gnu.org/licenses/>.

6. CITATION

Zhenguo Zhang and Daven C. Presgraves, CUA: a Flexible Codon Usage
Analyzer (In preparation)



Changes
Revision history for Bio-CUA

1.04    Thu May 28 10:37:37 EDT 2015
        1. Update Tutorial.pod by correcting format errors and reword
        sentences.
		2. Change program names as follows:
		   build_cai_param.pl -> cai_codon.pl
		   build_tai_param.pl -> tai_codon.pl
		   calculate_CUB.pl   -> cub_seq.pl
		The new names reflect the functions of the programs better.

1.03    Wed May 20 13:25:29 EDT 2015
        1. Add option --fop-param to calculate_CUB.pl so that the
		program can compute the metric Fop.
		2. Reformat the 'usage' information of the scripts in the
		'bin' directory.
		3. Add a tutorial file Tutorial.pod.

1.02    Sat May 16 23:04:25 EDT 2015
        Use a perl module t/env.pm to set environment variables during
		installation test, replacing the original way of setting
		environment variables using 'export' in makefile, because the
		latter does not work for NetBSD/FreeBSD/Windows systems

1.01    Thu Apr 30 16:51:09 EDT 2015
        First version, including methods to calculate codon usage bias
		indice at both codon and sequence levels.



t/generate_expected_data.pl
#!/usr/bin/env perl

use strict;
use lib '/home/zzhang65/Scripts/Perl/CPAN_dist/Bio-CUA/lib';
use Bio::CUA::CodonTable;
use Bio::CUA::Summarizer;
use Bio::CUA::SeqIO;
use Bio::CUA::CUB::Calculator;
use Bio::CUA::CUB::Builder;

warn "The program is to generate results for test purpose\n";
warn "Shall we continue?[Y/N]\n";

my $in = <STDIN>;
unless($in =~ /^Y/i)
{
	warn "Stopped\n";
	exit 0;
}

my $seqFile = 'test.fa'; # sequence data
my $tRNAFile = 'dmel_r5_apr2006.tRNA_copy'; # tRNA copy number

#my $outFile = shift || '-';

my $builder = Bio::CUA::CUB::Builder->new( -codon_table => 1 );

warn "# Generating expected results for Bio-CUA\n";

# 1. codon table part
my $codonTable = $builder->codon_table;
my $sect = 'Bio::CUA::CodonTable';
print ">>$sect\n";
print ">>>codon_to_AA\n";
my $codon2AA = $codonTable->codon_to_AA_map or die $!;
foreach my $c (sort keys(%$codon2AA))
{
	printf("%s\t%s\n", $c, $codon2AA->{$c});
}
print ">>>degeneracy\n";
my $codonDegeneracy = $codonTable->codon_degeneracy or die $!;
foreach my $deg (sort keys %$codonDegeneracy)
{
	foreach my $aa (keys %{$codonDegeneracy->{$deg}})
	{
		my $codons = $codonDegeneracy->{$deg}->{$aa};
		printf("%d\t%s\t%s\n", $deg, $aa, join(',', @$codons));
	}
}
print "<<$sect\n\n";


# 2. SeqIO part
$sect = 'Bio::CUA::SeqIO';
my $io = $sect->new(-file => $seqFile);
print ">>$sect\n";
print ">>>length\n";
while(my $seq = $io->next_seq)
{
	printf("%s\t%d\n", $seq->id, $seq->length);
}
print "<<$sect\n\n";

# 3. summarizer part
$sect = 'Bio::CUA::Summarizer';
my $sum = $sect->new(-codon_table => 1);
my $codonList = $sum->tabulate_codons($seqFile) or die $!;
print ">>$sect\n";
print ">>>codon_count\n";
_write_hash($codonList);
print "<<$sect\n\n";

# 4. builder part
$sect = 'Bio::CUA::CUB::Builder';
# traditional method, average, and background
my $rscu = $builder->build_rscu($seqFile,5,0.5); 
my $caiMax  = $builder->build_cai($seqFile,5,'max'); 
my $caiMean = $builder->build_cai($seqFile,5,'mean');
my $caiBack = $builder->build_b_cai($seqFile,$seqFile,5);
my $tai = $builder->build_tai($tRNAFile);
print ">>$sect\n";
print ">>>rscu\n";
_write_hash($rscu);
print ">>>cai_max\n";
_write_hash($caiMax);
print ">>>cai_mean\n";
_write_hash($caiMean);
print ">>>cai_back\n";
_write_hash($caiBack);
print ">>>tai\n";
_write_hash($tai);
print "<<$sect\n\n";

# 5. calculator part
$sect = 'Bio::CUA::CUB::Calculator';
my $calc = $sect->new(
	          -codon_table => 1,
			  -CAI_values  => $caiMax,
			  -tAI_values  => $tai
		  );

# calculate CAI/tAI/ENC/ENC_r for each sequence
$io = Bio::CUA::SeqIO->new(-file => $seqFile);
my %seqCAIs;
my %seqtAIs;
my %seqENCs;
my %seqENC_rs;

while(my $seq = $io->next_seq)
{
	$seqCAIs{$seq->id} = $calc->cai($seq);
	$seqtAIs{$seq->id} = $calc->tai($seq);
	$seqENCs{$seq->id} = $calc->enc($seq,5);
	$seqENC_rs{$seq->id} = $calc->enc_r($seq,5);
}

print ">>$sect\n";
print ">>>seq_cai\n";
_write_hash(\%seqCAIs);
print ">>>seq_tai\n";
_write_hash(\%seqtAIs);
print ">>>seq_enc\n";
_write_hash(\%seqENCs);
print ">>>seq_enc_r\n";
_write_hash(\%seqENC_rs);
print "<<$sect\n\n";


warn "Work done!!\n";

exit 0;

sub _write_hash
{
	my $hashRef = shift;
	foreach (sort keys(%$hashRef))
	{
		printf("%s\t%s\n",$_, $hashRef->{$_});
	}

	return 1;
}



lib/Bio/CUA.pm
package Bio::CUA;

use 5.006;
use strict;
use warnings;
use Carp;

# some global variables
our $VERSION = '1.03';
my $sep = "\t";
#my @openFHs; # all file handles opened by this class

=pod

=head1 NAME

Bio::CUA - Codon Usage Analyzer.

=head1 VERSION

Version 1.03

=head1 SYNOPSIS

This is the root class for the whole distribution of
L<http://search.cpan.org/dist/Bio-CUA/>,
providing some routine methods used by all classs in the
distribution L<http://search.cpan.org/dist/Bio-CUA/>. Users should not use this class
directly. Please start with its child classes such as
L<Bio::CUA::Summarizer>, L<Bio::CUA::CUB::Builder>.

=head1 DESCRIPTION

The aim of this distribution is to provide comprehensive and flexible
tools to analyze codon usage bias (CUB) and relevant problems, so that
users can speed up the genetic research by taking advantage of this
convenience.

One amino acid can be encoded by more than one synonymous codon, and
synonymous codons are unevenly used. For example, some codons are used
more often than other synonymous ones in highly expressed genes (I<Sharp
and Li 1987>). To measure the unevenness of codon usage, multiple
metrics of codon usage bias have been developed, such as Fop
(Frequency of optimal codons), CAI (Codon Adaptation Index), tAI (tRNA
Adaptation Index), and ENC (Effective Number of Codons). The causes of
CUB phenomena are complicated, including, mutational bias, selection on 
translational efficiency or accurancy. CUB is one fundamental concept
in genetics. 

So far, no software exists to compute all the above CUB metrics, and
more importantly parameters of CUB calculations are often fixed in
software, so one can only analyze genes in a limited list of species
and one can not incorporate its own parameters such as sequences of
highly expressed genes in a tissue. 

This package mainly solves these two problems. We also extend some
methods, such as GC-content corrected ENC, background-data normalized
CAI, etc. See the relevant methods in CUB classes for more details.

=head1 METHODS

=cut

sub new
{
	my ($caller, @args) = @_;
	my $self = {};
	my $class = ref($caller)? ref($caller) : $caller;

	bless $self, $class;
	my $hashRef = $self->_array_to_hash(\@args);

	# only process its own argument
	$self->debug(1) if($hashRef->{'debug'});

	return $self;
}

# store and retrieve tag values
sub get_tag
{
	my ($self, $tag) = @_;
	return $self->{'_tags'}->{$tag};
}

sub set_tag
{
	my ($self, $tag, $val) = @_;
	$self->{'_tags'}->{$tag} = $val;
}

=head2 debug

 Title   : debug
 Usage   : $true_of_false=$self->debug([$bool]);
 Function: get/set the boolean value.
 Returns : 0 as false, 1 as true
 Args    : optional. 0 or 1 for false and true, respectively.

=cut

sub debug
{
	my ($self, $val) = @_;
	$self->set_tag('debug', $val) if($val);
	return $self->get_tag('debug');
}


=head2 throw

 Title   : throw
 Usage   : $self->throw("Some fatal errors");
 Function: stop and report when fatal errors in formatted message
 Returns : None
 Args    : error message

=cut

# simplified version
sub throw
{
	my ($self, @args) = @_;
	my $class = ref($self) || $self;
	$class = ' '.$class if $class;
	my $title = "------------- EXCEPTION$class -------------";
	my $footer = ('-' x length($title))."\n";
	#my $text = join("\n", @args);
	my $text = _format_text(join(' ',@args));
	croak "\n$title\n", "MSG: $text\n", $footer, "\n";
}

=head2 warn

 Title   : warn
 Usage   : $self->warn("Please pay attention here")
 Function: report warning message when something looks not good
 Returns : None
 Args    : warning messages.

=cut

sub warn
{
	my ($self, @args) = @_;

	my $class = ref($self) || $self;
	$class = ' '.$class if $class;
	my $title = "------------- WARNING$class -------------";
	my $footer = ('-' x length($title))."\n";
	my $text = _format_text(join(' ',@args));
	#my $text = join("\n", @args);
	carp "\n$title\n", "MSG: $text\n", $footer, "\n";
}

# format the text into blocks with same line length
sub _format_text
{
	my ($text, $lineLen) = @_;

	$lineLen ||= 60;
	chomp($text);
	my $result = '';

	my @blocks = split /\n/, $text;

	foreach my $b (@blocks)
	{
		my $newB = _break_into_lines($b, $lineLen);
		$result .= $newB;
	}
	return $result;
}

sub _break_into_lines
{
	my ($text, $size) = @_;

	my $lines = '';
	my $textLen = length($text);

	my $accuLen = 0;
	while($accuLen < $textLen)
	{
		my $lineLen = $accuLen + $size > $textLen? $textLen - $accuLen
		: $size;
		my $l = substr($text,$accuLen,$lineLen);
		$accuLen += $lineLen;
		$lines .= $l."\n";
	}

	return $lines;
}

# return hash ref by reading into an array ref
sub _array_to_hash
{
	my ($self,$arrayRef,$nc) = @_;

	$self->throw("parameter '$arrayRef' to _array_to_hash is not an array reference")
	unless(ref($arrayRef) eq 'ARRAY');

	my %hash;

	$self->throw("Odd number of elements are in the array fed to",
		"_array_to_hash, check the array $arrayRef")
	unless($#$arrayRef % 2);

	for(my $i = 0; $i < $#$arrayRef; $i += 2)
	{
		my $k = $arrayRef->[$i];
		$k =~ s/^\-*//; # removing leading '-'
		$k = lc($k) unless($nc);
		$hash{$k} = $arrayRef->[$i+1];
	}

	return \%hash;
}


# write out hash to an outfile
sub _write_out_hash
{
	my ($self, $outFile, $hashRef) = @_;

	my $fh;
	open($fh, "> $outFile") or die "Can not open $outFile:$!";
	# let's sort the hash so that every time the same order is
	# produced
	my @sortedKeys = sort keys(%$hashRef);
	foreach my $k (@sortedKeys)
	{
		print $fh join($sep, $k, $hashRef->{$k}),"\n";
	}
	close $fh;

	return 1;
}

# open a file and return its file handle
sub _open_file
{
	my ($self, $file, $mode) = @_;

	$mode ||= ' ';

	my $fh;
	open($fh, "$mode $file") or $self->throw("can not open $file:$!");
	#push @openFHs, $fh;
	return $fh;
}

# parse the first $num fields of input file, and use the vaule at the
# first column as key
sub _parse_file
{
	my ($self, $file, $num) = @_;

	my %hash;
	my $fh = $self->_open_file($file);
	while(<$fh>)
	{
		next if /^#/ or /^\s*$/;
		chomp;
		s/^\s+//; # remove leading blanks
		my @fields = split /\s+/;
		if($num > 1)
		{
			$hash{uc($fields[0])} = $num > 2? 
				[@fields[1..($num-1)]] : $fields[1];
		}else
		{
			$hash{uc($fields[0])}++;
		}
	}
	close $fh;
	return \%hash;
}

# this method is called when object of this or child classes is being
# destroyed
# close the file handle if the object has one
sub DESTROY
{
	my $self = shift;
	#print $self;
	close $self->{'_fh'} if(exists $self->{'_fh'});
#	$self->SUPER::DESTROY(@_);
}

=head1 AUTHOR

Zhenguo Zhang, C<< <zhangz.sci at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-bio-cua at
rt.cpan.org> or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Bio-CUA>.  I will be
notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=cut

=head1 SUPPORT

You can find documentation for this class with the perldoc command.

	perldoc Bio::CUA

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Bio-CUA>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Bio-CUA>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Bio-CUA>

=item * Search CPAN

L<http://search.cpan.org/dist/Bio-CUA/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2015 Zhenguo Zhang.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see L<http://www.gnu.org/licenses/>.


=cut

1; # End of Bio::CUA



lib/Bio/CUA/Seq.pm
package Bio::CUA::Seq;

=head1 NAME

Bio::CUA::Seq - a module processing sequence object

=head1 SYNOPSIS

	use Bio::CUA::Seq;

	my $obj = Bio::CUA::Seq->new(
		-seq  => 'AGGTGCCG...',
		-id   => 'test_id',
		-desc => 'sequence description'
		);

	# available methods
	$obj->length; # get sequence length
	$obj->id; # get sequence id
	$obj->desc; # get sequence description
	$obj->seq; # get sequence string

=head1 DESCRIPTION

This module is called by L<Bio::CUA::SeqIO> to create sequence
object which has some basic methods required by the modules in the
distribution L<http://search.cpan.org/dist/Bio-CUA/>. The purpose of
this module is to increase the portability of the distribution.

=cut

use 5.006;
use strict;
use warnings;
use base qw/Bio::CUA/;

=head1 METHODS

=head2 new

 Title   : new
 Usage   : my $obj = Bio::CUA::Seq->new(%params);
 Function: create sequence object
 Returns : an object of this class
 Args    : arguments are specified in a hash and acceptable keys as follows:

=over 5

=item C<-seq>

the sequence string for the object, which can only be characters in
the range A-Z and a-z. Other characters are eliminated by the method.

=item C<-id>

the sequence name

=item C<-desc>

extra description of this sequence

=cut

sub new
{
	my ($caller, @args) = @_;

	my $self = $caller->SUPER::new(@args);

	my $hashRef = $self->_array_to_hash(\@args);

	my $str = $hashRef->{'seq'};
	$str =~ s/[^a-z]+//gi; # eliminate unknown chars

	$self->warn("No valid sequence string found",
	exists $hashRef->{'id'}? "for ".$hashRef->{'id'}:"")
	unless($str);

	$self->set_tag('seqstr', $str);
	$self->set_tag('id', $hashRef->{'id'}) if(exists $hashRef->{'id'});
	$self->set_tag('desc', $hashRef->{'desc'}) if(exists
		$hashRef->{'desc'});
	
	return $self;
}

=head2 length

 Title   : length
 Usage   : my $len = $self->length;
 Function: get the length of the sequence
 Returns : an integer
 Args    : None

=cut

sub length
{
	my $str = $_[0]->get_tag('seqstr');
	return length($str);
}

=head2 id

 Title   : id
 Usage   : my $id = $self->id;
 Function: get sequence name/id
 Returns : a string
 Args    : None

=cut

sub id
{
	$_[0]->get_tag('id');
}

=head2 desc

 Title   : desc
 Usage   : my $desc = $self->desc
 Function: get sequence description
 Returns : a string or undef if not exist
 Args    : None

=cut

sub desc
{
	$_[0]->get_tag('desc');
}

=head2 seq

 Title   : seq
 Usage   : my $seq = $self->seq
 Function: get sequence string of the object
 Returns : a string
 Args    : None

=cut

sub seq
{
	$_[0]->get_tag('seqstr');
}

=head1 AUTHOR

Zhenguo Zhang, C<< <zhangz.sci at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-bio-cua at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Bio-CUA>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Bio::CUA::Seq


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Bio-CUA>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Bio-CUA>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Bio-CUA>

=item * Search CPAN

L<http://search.cpan.org/dist/Bio-CUA/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2015 Zhenguo Zhang.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see L<http://www.gnu.org/licenses/>.


=cut

1; # End of Bio::CUA::Seq


Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;
use File::Spec;

WriteMakefile(
    NAME             => 'Bio::CUA',
    AUTHOR           => q{Zhenguo Zhang <zhangz.sci@gmail.com>},
    VERSION_FROM     => 'lib/Bio/CUA.pm',
    ABSTRACT_FROM    => 'lib/Bio/CUA.pm',
    LICENSE          => 'gpl_3',
	#PL_FILES         => {'other.PL' => ''}, # files run during compilation 
	EXE_FILES        => ["bin/cai_codon.pl",
	                     "bin/tai_codon.pl",
						 "bin/cub_seq.pl"
	                    ], # files in bin directory
    MIN_PERL_VERSION => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        #'ABC'              => 1.6,
        #'Foo::Bar::Module' => 5.0401,
		'File::Sort'        => 1.01,
		'File::Temp'        => 0,
		'Fcntl'             => 0,
		'parent'            => 0,
		'Getopt::Long'      => 0,
		'Carp'              => 0
    },

	# add more information
	META_MERGE  => {
		resources => {
			homepage => 'http://molevol.altervista.org/'
		}
	},

    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Bio-CUA-*' },
	test  => { RECURSIVE_TEST_FILES => 1, # test directories in t recursively
		# TESTS => 't/01-seqio.t'
	},
	# environment variables
	macro => {
#		'expected_file' => 't/expected.tsv' # expected results to compare
	}
);

package MY; # other content MakeMaker can't produce

sub postamble
{
	# The return will be directly write to makefile

	# get the command to set environment variables
#	my $os = $^O;
#	my $set = $os =~ /MSWin/i? 'set' : 'export';
#	my $expFile  = File::Spec->catfile('t', 'expected.tsv');
#	my $seqFile  = File::Spec->catfile('t', 'test.fa');
#	my $tRNAFile = File::Spec->catfile('t', 'dmel_r5_apr2006.tRNA_copy');
#
#	my $postamble = <<POST;
#export expected_file = $expFile
#export seq_file = $seqFile
#export tRNA_file = $tRNAFile
#POST

#	return $postamble;
}

# put subroutines here
